src/action/finder/hash.rs

```diff
@@ -0,0 +1,201 @@
+use crate::action::finder::request::HashActionOptions;
+use crate::fs::Entry;
+use crypto::digest::Digest;
+use log::warn;
+use rrg_macro::ack;
+use rrg_proto::file_finder_hash_action_options::OversizedFilePolicy;
+use rrg_proto::Hash as HashEntry;
+use std::cmp::min;
+use std::fs::File;
+use std::io::Read;
+use std::iter::repeat;
+
+/// Hashes data writen to it using SHA-1, SHA-256 and MD5 algorithms.
+struct Hasher {
+    /// Digest with SHA-1 hash.
+    sha1: crypto::sha1::Sha1,
+    /// Digest with SHA-256 hash.
+    sha256: crypto::sha2::Sha256,
+    /// Digest with MD5 hash.
+    md5: crypto::md5::Md5,
+    /// Stores total number of bytes inserted into hasher.
+    total_byte_count: u64,
+}
+
+impl Hasher {
+    pub fn new() -> Hasher {
+        Hasher {
+            sha1: crypto::sha1::Sha1::new(),
+            sha256: crypto::sha2::Sha256::new(),
+            md5: crypto::md5::Md5::new(),
+            total_byte_count: 0,
+        }
+    }
+}
+
+impl std::io::Write for Hasher {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.sha1.input(buf);
+        self.sha256.input(buf);
+        self.md5.input(buf);
+        self.total_byte_count += buf.len() as u64;
+
+        Ok(buf.len())
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        Ok(())
+    }
+}
+
+/// Performs `hash` action on the file in `entry` and returns the result to be reported in case of success.
+pub fn hash(entry: &Entry, config: &HashActionOptions) -> Option<HashEntry> {
+    match config.oversized_file_policy {
+        OversizedFilePolicy::Skip => {
+            if entry.metadata.len() > config.max_size {
+                return None;
+            }
+        }
+        OversizedFilePolicy::HashTruncated => {}
+    };
+
+    let file = ack! {
+        File::open(&entry.path),
+        error: "failed to open file: {}", entry.path.display()
+    }?;
+    let mut file = file.take(config.max_size);
```

Comment:
You can just chain `take`, no need to introduce a variable only to shadow it on the next line.

Reply:
I believe `ack! {...}?.take()` is a bit too much for 1 statement. If you disagree I'm happy to change it.